qa/rpc-tests/sendheaders.py

```diff
@@ -0,0 +1,425 @@
+
+    def run_test(self):
+        # Setup the p2p connections and start up the network thread.
+        inv_node = InvNode()
+        test_node = TestNode()
+
+        connections = []
+        connections.append(NodeConn('127.0.0.1', p2p_port(0), self.nodes[0], inv_node))
+        connections.append(NodeConn('127.0.0.1', p2p_port(0), self.nodes[0], test_node))
+        inv_node.add_connection(connections[0])
+        test_node.add_connection(connections[1])
+
+        NetworkThread().start() # Start up network handling in another thread
+
+        # Test logic begins here
+        inv_node.wait_for_verack()
+        test_node.wait_for_verack()
+
+        tip = int("0x" + self.nodes[0].getbestblockhash() + "L", 0)
+
+        # PART 1
+        # 1. Mine a block; expect inv announcements each time
+        print "Part 1: headers don't start before sendheaders message..."
+        for i in xrange(4):
+            old_tip = tip
+            tip = self.mine_blocks(1)
+            assert_equal(inv_node.check_last_announcement(inv=[tip]), True)
+            assert_equal(test_node.check_last_announcement(inv=[tip]), True)
+            # Try a few different responses; none should affect next announcement
+            if i == 0:
+                # first request the block
+                test_node.get_data([tip])
+                test_node.wait_for_block(tip, timeout=5)
+            elif i == 1:
+                # next try requesting header and block
+                test_node.get_headers(locator=[old_tip], hashstop=tip)
+                test_node.get_data([tip])
+                test_node.wait_for_block(tip)
+                test_node.clear_last_announcement() # since we requested headers...
+            elif i == 2:
+                # this time announce own block via headers
+                height = self.nodes[0].getblockcount()
+                last_time = self.nodes[0].getblock(self.nodes[0].getbestblockhash())['time']
+                block_time = last_time + 1
+                new_block = create_block(tip, create_coinbase(height+1), block_time)
+                new_block.solve()
+                test_node.send_header_for_blocks([new_block])
+                test_node.wait_for_getdata([new_block.sha256], timeout=5)
+                test_node.send_message(msg_block(new_block))
+                test_node.sync_with_ping() # make sure this block is processed
+                inv_node.clear_last_announcement()
+                test_node.clear_last_announcement()
+
+        print "Part 1: success!"
+        print "Part 2: announce blocks with headers after sendheaders message..."
+        # PART 2
+        # 2. Send a sendheaders message and test that headers announcements
+        # commence and keep working.
+        test_node.send_message(msg_sendheaders())
+        prev_tip = int("0x" + self.nodes[0].getbestblockhash() + "L", 0)
+        test_node.get_headers(locator=[prev_tip], hashstop=0L)
```

Comment:
Surprisingly, this code is necessary to make the test work.  Turns out there's an unusual feature in the getheaders handling, where we set `pindexBestHeaderSent` to `chainActive.Tip()` in two situations: if we actually are sending the tip, or if we have nothing to send.  I think the latter case can only happen if our peer has our tip in their locator, so I believe this behavior is desirable, if surprising.

I've added a comment to the relevant line of code in main.cpp explaining this behavior.